internal/resolution/client/maven_registry_client.go

```diff
@@ -0,0 +1,249 @@
+	if vk.System != resolve.Maven {
+		return nil, fmt.Errorf("wrong system: %v", vk.System)
+	}
+
+	g, a, found := strings.Cut(vk.Name, ":")
+	if !found {
+		return nil, fmt.Errorf("invalid Maven package name %s", vk.Name)
+	}
+	proj, err := c.api.GetProject(ctx, g, a, vk.Version)
+	if err != nil {
+		return nil, err
+	}
+
+	// Only merge default profiles by passing empty JDK and OS information.
+	if err := proj.MergeProfiles("", maven.ActivationOS{}); err != nil {
+		return nil, err
+	}
+	// We need to merge parents for potential dependencies in parents.
+	if err := mavenutil.MergeParents(ctx, c.api, &proj, proj.Parent, 1, "", false); err != nil {
+		return nil, err
+	}
+	proj.ProcessDependencies(func(groupID, artifactID, version maven.String) (maven.DependencyManagement, error) {
+		root := maven.Parent{ProjectKey: maven.ProjectKey{GroupID: groupID, ArtifactID: artifactID, Version: version}}
+		var result maven.Project
+		if err := mavenutil.MergeParents(ctx, c.api, &result, root, 0, "", false); err != nil {
+			return maven.DependencyManagement{}, err
+		}
+
+		return result.DependencyManagement, nil
+	})
+
+	reqs := make([]resolve.RequirementVersion, 0, len(proj.Dependencies))
+	for _, d := range proj.Dependencies {
+		reqs = append(reqs, resolve.RequirementVersion{
+			VersionKey: resolve.VersionKey{
+				PackageKey: resolve.PackageKey{
+					System: resolve.Maven,
+					Name:   d.Name(),
+				},
+				VersionType: resolve.Requirement,
+				Version:     string(d.Version),
+			},
+			Type: resolve.MavenDepType(d, ""),
+		})
+	}
+
+	return reqs, nil
+}
+
+func (c *MavenRegistryClient) MatchingVersions(ctx context.Context, vk resolve.VersionKey) ([]resolve.Version, error) {
+	if vk.System != resolve.Maven {
+		return nil, fmt.Errorf("wrong system: %v", vk.System)
+	}
+
+	g, a, found := strings.Cut(vk.Name, ":")
+	if !found {
+		return nil, fmt.Errorf("invalid Maven package name %s", vk.Name)
+	}
+	metadata, err := c.api.GetArtifactMetadata(ctx, g, a)
+	if err != nil {
+		return nil, err
+	}
+
+	versions := make([]resolve.Version, len(metadata.Versioning.Versions))
+	for i, v := range metadata.Versioning.Versions {
+		versions[i] = resolve.Version{
+			VersionKey: resolve.VersionKey{
+				PackageKey:  vk.PackageKey,
+				Version:     string(v),
+				VersionType: resolve.Concrete,
+			},
+		}
+	}
+
+	return resolve.MatchRequirement(vk, versions), nil
+}
+
+func (c *MavenRegistryClient) WriteCache(path string) error {
+	f, err := os.Create(path + mavenRegistryCacheExt)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	return gob.NewEncoder(f).Encode(c.api)
+}
+
+func (c *MavenRegistryClient) LoadCache(path string) error {
+	f, err := os.Open(path + mavenRegistryCacheExt)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	return gob.NewDecoder(f).Decode(&c.api)
+}
+
+func (c *MavenRegistryClient) PreFetch(ctx context.Context, imports []resolve.RequirementVersion, manifestPath string) {
```

Comment:
For my own reference: I actually dislike needing to make `PreFetch` an interface method of each `DependencyClient`, since it's almost the same for each client, and also because it needs the deps.dev API to work.

After this is merged, I want to look into making `PreFetch` a standalone function that calls every `resolve.Client` method for caching, and use something like [singleflight](https://pkg.go.dev/golang.org/x/sync/singleflight) to avoid making identical network requests multiple times.